### PR TITLE
tests: Add capability to disable RPC timeout in functional tests

### PR DESCRIPTION
**Pull request by rajarshimaitra**
Many times, especially while debugging RPC callbacks to core using gdb, the test timeout kicks in before the response can get back. This can be annoying and requires restarting the functional test as well as gdb attachment. 

This PR adds a `--notimeout` flag into `test_framework` and sets the `rpc_timeout` accordingly if the flag is set.

The same effect can be achieved with newly added `--factor` flag but keeping a separate flag that explicitly disables the timeout can be easier for new testers to find it out and separates its purpose from the `--factor` flag.

Requesting review ryanofsky jnewbery as per the IRC discussion.

Update: After initial round of review, the approach is modified to accommodate the functionality in already existing `--factor` flag. `--factor` is changed to `--timeout-factor` to express its intent better.

### CODE REVIEW

File: test/functional/test_framework/test_framework.py [hunk 1]
```diff
@@ -227,6 +229,11 @@ def setup(self):
             if not self.supports_cli:
                 raise SkipTest("--usecli specified but test does not support using CLI")
             self.skip_if_no_cli()
+
+        #Diasble timeout if flag is set
+        if self.options.notimeout:
```
Comment (jonatack) [inline test/functional/test_framework/test_framework.py:234]: Should it warn, raise, or not be set if `--factor` is passed?

**Review by jonatack — COMMENTED** [review]
Concept ACK

**Comment by jonatack** [comment]
@rajarshimaitra: don't forget to add the prefix "test: " to your PR title (see `CONTRIBUTING.md` for more) :bulb:

File: test/functional/test_framework/test_framework.py [hunk 2]
```diff
@@ -227,6 +229,11 @@ def setup(self):
             if not self.supports_cli:
                 raise SkipTest("--usecli specified but test does not support using CLI")
             self.skip_if_no_cli()
+
+        #Diasble timeout if flag is set
```
[the remark this answers is not in the document]
Comment (rajarshimaitra) [reply test/functional/test_framework/test_framework.py:233]: Done.

File: test/functional/test_framework/test_framework.py (hunk 1 above)
Comment (rajarshimaitra) [reply test/functional/test_framework/test_framework.py:234]: I added a message in the help text. But yes, it would be a good idea to enforce it.

File: test/functional/test_framework/test_framework.py [hunk 3]
```diff
@@ -170,6 +170,8 @@ def parse_args(self):
         parser.add_argument("--descriptors", default=False, action="store_true",
                             help="Run test using a descriptor wallet")
         parser.add_argument('--factor', type=float, default=1.0, help='adjust test timeouts by a factor')
+        parser.add_argument("--notimeout", dest="notimeout", default=False, action="store_true",
+                            help="Disable the test timeout, should not be used with --factor")
```
Comment (MarcoFalke) [inline test/functional/test_framework/test_framework.py:174]: why not make any negative factor map to factor=inf?

File: test/functional/test_framework/test_framework.py (hunk 1 above)
Comment (jonatack) [reply test/functional/test_framework/test_framework.py:234]: I'm thinking either --notimeout should override any --factor setting and output a warning if it does, or halt with an explanation. Depending on which, the helps should be updated. Curious what others think.

File: test/functional/test_framework/test_framework.py (hunk 3 above)
Comment (MarcoFalke) [reply test/functional/test_framework/test_framework.py:174]: (ofc with proper documentation)

Comment (jonatack) [reply test/functional/test_framework/test_framework.py:174]: i like the simplicity of this

Comment (rajarshimaitra) [reply test/functional/test_framework/test_framework.py:174]: That definitely can be done, but I feel it would unnecessarily complicate it for new testers. It won't be readily obvious unless they look it up. 
Also, I have issues with factor doc, it wasn't clear that it related to rpc_timeout until I look it up. (should it be changed?) 
Lastly I feel keeping the --notimeout flag explicit, helps differentiate the purpose of two.

File: test/functional/test_framework/test_framework.py (hunk 1 above)
Comment (rajarshimaitra) [reply test/functional/test_framework/test_framework.py:234]: I think it's better to immediately raise an error stating these two should not be set together. 
Generally, someone wanting to disable timeout will not have a need to set --factor. Catching that would be sufficient.
But there can be an edge case where for some reason they need --factor but also want to attach gdb to the process. Right now that's not possible without much deeper refactoring. currently http responses in a test to node, and node to node, are checked in the same function . So it's not possible to only disable it for test to node, but keep it for node to node.

File: test/functional/test_framework/test_framework.py (hunk 3 above)
Comment (MarcoFalke) [reply test/functional/test_framework/test_framework.py:174]: >  it wasn't clear that it related to rpc_timeout 

then that sounds like a documentation issue that should be fixed, not worked around

Comment (rajarshimaitra) [reply test/functional/test_framework/test_framework.py:174]: I hope its clear that the work is not around --factor. I have stated that the similar effect can be achieved by just using a higher factor. The addition only helps to reduce the burden of figuring it out. There are people who are setting env variables and manually changing the code to get this behaviour. Having a flag only simplifies that issue.

**Comment by jnewbery** [comment]
I agree with @MarcoFalke. This functionality already exists with `--factor`. Rather than adding redundant code, it'd be better to improve the documentation, both in the help text and https://github.com/bitcoin/bitcoin/blob/master/test/README.md#troubleshooting-and-debugging-test-failures.

**Comment by rajarshimaitra** [comment]
> I agree with @MarcoFalke. This functionality already exists with `--factor`. Rather than adding redundant code, it'd be better to improve the documentation, both in the help text and https://github.com/bitcoin/bitcoin/blob/master/test/README.md#troubleshooting-and-debugging-test-failures.

Acknowledged. The factor doc can be improved and the debugging doc can be amended with fixing around this possible problem. On the general approach, is it better to disable timeout by setting a negative factor as @MarcoFalke suggested and amending the doc or people should just use a higher factor?

My personal view is, there should be an easy way to disable the timeout without with or without a new flag.         

Also regarding [this point](https://github.com/bitcoin/bitcoin/pull/18986#discussion_r426149845), is it (at least theoretically) possible to handle such edge cases?

**Comment by jnewbery** [comment]
>  is it better to disable timeout by setting a negative factor as @MarcoFalke suggested and amending the doc or people should just use a higher factor?

My two cents: setting `factor=0` should disable all timeouts.

I also think the `factor` option and variable name should be changed everywhere. The name factor by itself is almost meaningless. Perhaps _timeout_factor_?

**Comment by rajarshimaitra** [comment]
> My two cents: setting `factor=0` should disable all timeouts.

That with added doc should serve the purpose. But It also should handle other areas properly where factor value is being used. If everyone agrees, I can start implementing this.  

> I also think the `factor` option and variable name should be changed everywhere. The name factor by itself is almost meaningless. Perhaps _timeout_factor_?

Precisely my thought. The name doesn't disclose the intent. _timout_factor_ would be more revealing even if a little long.

Curious what others think.

**Comment by jonatack** [comment]
> > is it better to disable timeout by setting a negative factor as @MarcoFalke suggested and amending the doc or people should just use a higher factor?
> 
> My two cents: setting `factor=0` should disable all timeouts.
> 
> I also think the `factor` option and variable name should be changed everywhere. The name factor by itself is almost meaningless. Perhaps _timeout_factor_?

SGTM.

**Comment by jonatack** [comment]
> My two cents: setting `factor=0` should disable all timeouts.
> 
> I also think the `factor` option and variable name should be changed everywhere. The name factor by itself is almost meaningless. Perhaps _timeout_factor_?

SGTM

**Comment by rajarshimaitra** [comment]
Thanks everyone for all the inputs. I have modified the feature accordingly.

- `--factor` is changed to `--timeout-factor` to better express intent and help doc updated. Internally all `self.options.factor` modified to `self.options.timeout_factor`.

- setting `--timeout-factor 0` will disable RPC timeout for functional tests. Fixed few typos as I was touching these files anyway.

- [/test/README.md](https://github.com/rajarshimaitra/bitcoin/tree/notimeout-flag/test#attaching-a-debugger) updated to add a note on how to handle potential RPC timeout while attaching gdb. (even though small change, to adhere to the general guideline, its made into a separate commit)

- Verified all existing tests are passing as it is.

- Verified all existing tests are passing with `--timeout-factor 0`.

I think this PR is now ready for a second round of review. Polite request to reviewers to take a second look.

File: test/functional/test_framework/test_framework.py [hunk 4]
```diff
@@ -102,7 +102,10 @@ def __init__(self):
         self.bind_to_localhost_only = True
         self.set_test_params()
         self.parse_args()
-        self.rpc_timeout = int(self.rpc_timeout * self.options.factor) # optionally, increase timeout by a factor
+        if self.options.timeout_factor == 0 :
+            self.rpc_timeout = 99999
```
Comment (MarcoFalke) [inline test/functional/test_framework/test_framework.py:106]:
```suggestion
            self.options.timeout_factor = 99999
```

This should apply to all timeouts, otherwise the tests will crash in `sync_*` calls

File: test/functional/test_framework/test_framework.py [hunk 5]
```diff
@@ -169,7 +172,7 @@ def parse_args(self):
                             help="set a random seed for deterministically reproducing a previous test run")
         parser.add_argument("--descriptors", default=False, action="store_true",
                             help="Run test using a descriptor wallet")
-        parser.add_argument('--factor', type=float, default=1.0, help='adjust test timeouts by a factor')
+        parser.add_argument('--timeout-factor', dest="timeout_factor", type=float, default=1.0, help='adjust RPC timeout by a factor. Setting it to 0 disables the timeout')
```
Comment (MarcoFalke) [inline test/functional/test_framework/test_framework.py:175]:
```suggestion
        parser.add_argument('--timeout-factor', dest="timeout_factor", type=float, default=1.0, help='adjust test timeouts by a factor. Setting it to 0 disables the timeout')
```

The previous wording was correct. This is not only about rpc timeouts

File: test/functional/test_framework/util.py [hunk 6]
```diff
@@ -208,10 +208,13 @@ def str_to_b64str(string):
 def satoshi_round(amount):
     return Decimal(amount).quantize(Decimal('0.00000001'), rounding=ROUND_DOWN)
 
-def wait_until(predicate, *, attempts=float('inf'), timeout=float('inf'), lock=None, factor=1.0):
-    if attempts == float('inf') and timeout == float('inf'):
+def wait_until(predicate, *, attempts=float('inf'), timeout=float('inf'), lock=None, timeout_factor=1.0):
+    if timeout_factor == 0:
```
Comment (MarcoFalke) [inline test/functional/test_framework/util.py:212]: Why is this needed? The condition should never be True.

File: test/functional/test_framework/test_framework.py (hunk 4 above)
Comment (rajarshimaitra) [reply test/functional/test_framework/test_framework.py:106]: Right, I missed that part. Fixed.

File: test/functional/test_framework/test_framework.py (hunk 5 above)
Comment (rajarshimaitra) [reply test/functional/test_framework/test_framework.py:175]: Reverted.

File: test/functional/test_framework/util.py (hunk 6 above)
Comment (rajarshimaitra) [reply test/functional/test_framework/util.py:212]: This was required because I was setting the factor to 0. After the above changes it's not required anymore. Reverted.

**Comment by rajarshimaitra** [comment]
Addressed @MarcoFalke 's comments.

**Comment by MarcoFalke** [comment]
ACK 38c3dd9c706e7e84b2a4dbaf1424a3f1c3b694fc and thanks for fixing up all my typos :sweat_smile:

File: test/functional/test_framework/util.py [hunk 7]
```diff
@@ -208,10 +208,10 @@ def str_to_b64str(string):
 def satoshi_round(amount):
     return Decimal(amount).quantize(Decimal('0.00000001'), rounding=ROUND_DOWN)
 
-def wait_until(predicate, *, attempts=float('inf'), timeout=float('inf'), lock=None, factor=1.0):
+def wait_until(predicate, *, attempts=float('inf'), timeout=float('inf'), lock=None, timeout_factor=1.0):
```
Comment (jnewbery) [inline test/functional/test_framework/util.py:211]: (for a follow-up PR) this is a really weird interface. The client is saying: 'take these two numbers, multiply them and use that as the timeout'. The client is able to do that multiplication itself! Why does it need to pass in two different values?

**Review by jnewbery — COMMENTED** [review]
ACK 38c3dd9c706e7e84b2a4dbaf1424a3f1c3b694fc.

Thanks for doing this @rajarshimaitra. I've added a suggestion for a follow-up PR if you're interested.

File: test/functional/test_framework/util.py (hunk 7 above)
Comment (rajarshimaitra) [reply test/functional/test_framework/util.py:211]: Thanks for the review. The process of making a PR was educational and made me go deeper into stuffs. Couldn't have done without your review club.
   
And also thanks for the follow-up idea. Yes it does seem redundant. Surely this can be improved. I will look into this. :+1:

Comment (rajarshimaitra) [reply test/functional/test_framework/util.py:211]: On an initial look it seems this is fairly trivial. There is no test function that calls `wait_until()` with `timeout_factor`. Either uses default or specifies a specific timeout. The only place `timeout_factor` is used is at node initialization and peer connection logic, which already knows how to handle it. 
So it seems it just requires changing the `wait_until()` signature.
Let me know if I am missing something there.

File: test/README.md [hunk 8]
```diff
@@ -225,6 +225,10 @@ gdb /home/example/bitcoind <pid>
 Note: gdb attach step may require ptrace_scope to be modified, or `sudo` preceding the `gdb`.
 See this link for considerations: https://www.kernel.org/doc/Documentation/security/Yama.txt
 
+Often while debugging rpc calls from functional tests, the test might reach timeout before
+process can return a response. Use `--timeout-factor 0` to disable all rpc timeouts for that partcular
```
Comment (MarcoFalke) [inline test/README.md:229]:
```suggestion
process can return a response. Use `--timeout-factor 0` to disable all rpc timeouts for that particular
```

**Review by maflcko — COMMENTED** [review]
Looks like there is a typo? Mind to fix it up in your follow-up pull request?
